onnxscript/optimizer/_inliner.py

```diff
@@ -0,0 +1,159 @@
+
+import onnxscript.ir as ir
+import onnxscript.ir.convenience as convenience
+
+# A replacement for a node specifies a list of nodes that replaces the original node,
+# and a list of values that replaces the original node's outputs.
+# If the replacement is None, it indicates that the node should not be replaced.
+
+NodeReplacement = Optional[Tuple[Iterable[ir.Node], Iterable[ir.Value]]]
+
+
+class CopyReplace:
+    """Utilities for creating a copy of IR objects with substitutions for attributes/input values."""
+
+    def __init__(
+        self, attr_map: dict[str, ir.Attr | ir.RefAttr], value_map: dict[ir.Value, ir.Value]
+    ) -> None:
+        self._value_map = value_map
+        self._attr_map = attr_map
+
+    def clone_value(self, value: ir.Value) -> ir.Value:
+        # Only input-values are cloned.
+        if value.producer() is not None:
+            raise ValueError("Only input values should be cloned.")
+        new_value = ir.Value(
+            name=value.name, type=value.type, shape=value.shape, doc_string=value.doc_string
+        )
+        self._value_map[value] = new_value
+        return new_value
+
+    def clone_attr(self, key: str, attr: ir.Attr | ir.RefAttr) -> ir.Attr | ir.RefAttr | None:
+        if isinstance(attr, ir.Attr):
+            if attr.type == ir.AttributeType.GRAPH:
+                return self.clone_graph(attr.value)
+            elif attr.type == ir.AttributeType.GRAPHS:
+                return [self.clone_graph(graph) for graph in attr.value]
+            return attr
+        assert isinstance(attr, ir.RefAttr)
+        if key in self._attr_map:
+            return self._attr_map[key]
+        # Note that if a function has an attribute-parameter X, and a call (node) to the function
+        # has no attribute X, all references to X in nodes inside the function body will be
+        # removed. This is just the ONNX representation of optional-attributes.
+        return None
+
+    def clone_node(self, copied: ir.Node) -> ir.Node:
+        new_inputs = [self._value_map[input] for input in copied.inputs]
+        new_attributes = [
+            new_value
+            for key, value in copied.attributes.items()
+            if (new_value := self.clone_attr(key, value)) is not None
+        ]
+        new_node = ir.Node(
+            copied.domain,
+            copied.op_type,
+            new_inputs,
+            new_attributes,
+            overload=copied.overload,
+            num_outputs=len(copied.outputs),
+            graph=None,  #  TODO:
+            name=copied.name,  #  TODO: add a unique name
+            doc_string=copied.doc_string,
+            metadata_props=copied.metadata_props,
+        )
+        new_outputs = new_node.outputs
+        for i, output in enumerate(copied.outputs):
+            self._value_map[output] = new_outputs[i]
+            new_outputs[i].name = output.name
+        return new_node
+
+    def clone_graph(self, graph: ir.Graph) -> ir.Graph:
```

Comment:
Thanks! I wonder if we should implement a `copy()` method in Graph, or implement the `__deepcopy__` magic for it to provide pythonic apis?

Reply:
It is not similar to standard copy, though. It is dependent on some state/parameters (mappings for outer-scope references inside the graph/node).

Reply:
What is done here is not suitable for copy/clone and vice-versa.  Different logic. At best, we could rename these methods and call them by another name like "apply_substitutions" or "apply_replacements". I thought that the name of the class "CopyReplace" would suffice to indicate what's meant, but it can be improved.

Reply:
I see. Thanks!